jax/interpreters/xla.py

```diff
@@ -601,15 +601,18 @@ def _xla_callable_device(nreps, backend, device, arg_devices):
     else:
       assert False  # Unreachable given the error check in _xla_callable
 
-def _xla_callable_args(c, avals, tuple_args):
+def _xla_callable_args(c, avals, tuple_args, replicated=None):
   if not tuple_args:
-    xla_args = [xb.parameter(c, i, aval_to_xla_shape(a))
+    if replicated is None:
+      replicated = [None] * len(avals)
+    xla_args = [xb.parameter(c, i, aval_to_xla_shape(a), replicated=r)
                 if a is not abstract_token else xops.CreateToken(c)
-                for i, a in enumerate(avals)]
+                for i, (a, r) in enumerate(zip(avals, replicated))]
     return xla_args
   else:
-    tuple_param = xb.parameter(c, 0, xc.Shape.tuple_shape(
-        [aval_to_xla_shape(a) for a in avals if a is not abstract_token]))
+    tuple_shape = xc.Shape.tuple_shape(
+        [aval_to_xla_shape(a) for a in avals if a is not abstract_token])
+    tuple_param = xb.parameter(c, 0, tuple_shape, replicated=replicated)
```

Comment:
There is some dependency between the shape of `replicated` and which args are tokens. I suspect that this code does not work properly in presence of token arguments. That probably deserves its own investigation.

Reply:
Are token arguments allowed at the top level? Non-`None` replication annotations are forbidden for subcomputations, so it's possible that they never intersect.

Reply:
Yes, they are (in JAX, but not XLA): https://github.com/google/jax/commit/4aaf325d3a16a12d3db5f84af51103e4f07a3374#diff-bdad3557763a5ef51aada454f6c2e496L245-L246
So I do need to deal with this.